### PR TITLE
feat(flyout): initial release of flyout module

### PR DESCRIPTION
**Pull request by prudho**
## Description
The following PR implements a new module called `Flyout`. It's based on my own ES6 fomantic code, which is derived from `sidebar`, and CSS code is based from @jamessampford #1885 tryout.

For now this is just an embryo, and we need to discuss further to decide how we implement it.

## Testcase
[JSFiddle](https://jsfiddle.net/4wbct2ns/)

## Screenshot
Flyout

## Related issues
#121, #1885

### CODE REVIEW

**Comment by lubber-de** [comment]
Looks very promising! 👍🏼 
I change this PR into Draft, if you don't mind :) At least some CSS adjustments are needed from a first quick 
check. I will test this more deeper later

**Comment by prudho** [comment]
Yeah i'm not very fluent with CSS yet...

Here are some of my thoughts about module:
- `overlay` should be the default transition. Maybe we should remove the push transition, code would be more clear and I don't think that a flyout should push page content.
- dealing with width should be different from sidebar. IMHO, the width classes `thin` and `wide` are not really adapted. Maybe we could use a grid system instead, like `ui eight wide flyout` dimensionned from `context` width.
- Mobile handling: the @jamessampford's idea for a `fullscreen` parameter is cool, and should be a default behavior.
- Adding a pure js implementation like we did for modals seems to be a great idea.

**Comment by testinfected** [comment]
Hi guys, what's the status on this. Our team is very much interested

Thanks

**Comment by prudho** [comment]
Hi @testinfected. This is still a work in progress for now, as I barely touched it since July. But it should be released in 2.9 (let's say before the end of the year). I'll push some commits later.

**Comment by lubber-de** [comment]
I played with this now, here are my thoughts (not meant as a complete review as it is declared as WIP)

### Overall
I havent really compared all the code, but i feel that most of the code is the same as for sidebar, so i thought if there is a possibility to just merge them and make flyout a variant of sidebar to have a smaller codebase. 🤔 

### Quick feedback
- test, if it works together with sidebar (using same pusher element) 
- prepare more examples in a jsfiddle to play with/show all the features/variants (will also be a good base for the later docs page)
- fix the vertical scrollbar for the whole (left/right) flyout appearing all the time
- define "multiple", so a previous flyout gets a dedicated inner dimmer (like modals have)
- create flyouts via pure js properties (you already mentioend that above)
- put more of css values from .less to .variables to make them customizable
- the "x wide flyout" approach (like grid) seems to be a nice idea at first
	- however that would mean the width is always a percentage  
	- Especially on mobile devices the "x wide" approach wont work well
	- it is not needed at all when using "top/bottom" flyouts
	- So i think we might implement this only in _addition_ to "very/wide/thin" 
- when i use `flyout.flyout({}).flyout('toggle');` ( to be able to provide custom settings to the flyout) the second flyout in the jsfiddle is not shown anymore. Instead, the first flyout just closes.
- veryThinWidth: 60px seems very less for what a flyout is supposed to be (an alternative to modal). 60px would be more like a classical quick icons menu showing up (which we already have sidebar for). Of course, I know this value is customizable :wink:

**Comment by jamessampford** [comment]
If merged with sidebar, would need to consider if can offer different default settings for a flyout and some of the separate flyout-specific settings could possibly also cause confusion (though docs could possibly point it out)

That said, I wonder if shared functionality could be bundled into some kind of base class with both sidebar and flyout extended from the base class (or if flyout could extend/override sidebar)?

it may be easier if just overlay and not push, but then loses some flexibility

With regards to mobile, my original thinking was to have a setting to force mobile to fill max width … an alternative could be to use attributes, eg. `data-width=“200px” data-mobile=“100%”` with something similar in place for height for top/bottom (which could possibly allow for mobile menus) … think `x wide` just tries to bring some consistency, isn’t there also something like `x wide y wide mobile`?

**Review by lubber-de — COMMENTED** [review]
I just did a quick pre-review. That's why this review is just a comment instead of "requesting changes" for now.
I already said previously to use LESS variables instead of hardcoded/compiled CSS values, so i did not create comments for each of those sections (would be too many)😉 .  
I quickly tested if all those !important statements are really necessary and they don't, so please remove them.
For the wide variants we already have LESS variables used by form, table and grid, so flyout should use them as well.

If you don't want me to do further reviews until you set the PR from Draft to "ready for review" please tell me and i'll stay quiet 😆 
Please also update the jsfiddle to use the latest code then (and all possible variants). Thx.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 34a79c9579532c21c511462024cd5f7b859780ad.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 9cab8607332638da732c059180a79878688d2ee3.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 9cab8607332638da732c059180a79878688d2ee3.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 9cab8607332638da732c059180a79878688d2ee3.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 9cab8607332638da732c059180a79878688d2ee3.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 9cab8607332638da732c059180a79878688d2ee3.

**Comment by prudho** [comment]
Thanks @lubber-de ! Don't hesitate to suggest me changes to make the code better, as I already said CSS is a pain for me :)

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Fixed by 9cab8607332638da732c059180a79878688d2ee3.

**Comment by prudho** [comment]
I've updated the [JSFiddle](https://jsfiddle.net/yxmok5eb/) with added scenarios. Looks like I have some errors on content height that I need to fix before going further...

**Comment by prudho** [comment]
Update [JSFiddle](https://jsfiddle.net/4wbct2ns/) again, whith more example (very/thin/wide), a fix for content height in specific context and increase width values.

I think that, in term of functionalities, module can be proposed as it in 2.9. Some clean up and hardcoded variables to eliminate and it can be shipped. What do you think @fomantic/maintainers ?

**Comment by lubber-de** [comment]
@prudho We should at least also implement
> create flyouts via pure js properties
before releasing this , as we now already have that feature for modal and toast and would be consistent.

You might adopt that from modal as it also has a11y support, tabbing support, close focus, etc. (and infact flyout is a visual combo of modal and sidebar)

File: src/definitions/modules/flyout.less
Comment (lubber-de) [inline src/definitions/modules/flyout.less:622]: This is independent of the positioning as it is always the same transform value, so this can be reduced to
```suggestion
  /* End - flyout */
  .ui.visible.overlay.flyout {
    transform: translate3d(0%, 0, 0);
  }
```

File: src/definitions/modules/flyout.less
Comment (lubber-de) [inline src/definitions/modules/flyout.less:631]: You removed everything but `overlay` in the JS code by 4ebcf020dbf2da8839e9e61a7bbc7399deae5100, so there is no support for `push`, `uncover`, `slide along`, `slide out` or `scale down` anymore.
Please remove all of the related LESS code for those unsupported transitions as well.

File: src/themes/default/globals/variation.variables
Comment (lubber-de) [inline src/themes/default/globals/variation.variables:566]: You removed everything but `overlay` in the JS code by 4ebcf020dbf2da8839e9e61a7bbc7399deae5100, so there is no support for `push`, `uncover`, `slide along`, `slide out` or `scale` anymore.
Please remove all of the related variables for those unsupported transitions as well.

**Review by lubber-de — CHANGES_REQUESTED** [review]
Please see my comments for optimizing the LESS code.
Regarding the JS code i commented about missing aria, focus-trap, etc in https://github.com/fomantic/Fomantic-UI/pull/2004#issuecomment-1136215527
As this also needs a whole dedicated docs page, i, for now, milestone this to 2.9.x

**Comment by prudho** [comment]
@lubber-de thanks for the feedback. Yes I still plan it to be released in 2.9... I miss time actually but i'll review your requested changes and create the doc ASAP.

For me the module can be released even if all features are not implemented or optimized yet. Feedback from users is always a good sight for what can be fixed and patches like 2.9.1 and so on will be perfect for that.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less:480]: It's to prevent an eventual user error:
image

This can happen if user use `ui top eight wide flyout`, for example. But if you think it is unnecessary, i'll remove it.

Comment (lubber-de) [reply src/definitions/modules/flyout.less:480]: I still think we should remove/simply that. A `wide` flyout is supposed to hava a dedicated width (which does not really make sense if one also uses "top", but if one still combines it, then the screenshot is the expected result (the flyout is on top and is eight wide ;))

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Resolved by deb1a561dfc99c26494d0e5c503c1118c7d4ccd3.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less:42]: Resolved by deb1a561dfc99c26494d0e5c503c1118c7d4ccd3.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Resolved by deb1a561dfc99c26494d0e5c503c1118c7d4ccd3.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Resolved by deb1a561dfc99c26494d0e5c503c1118c7d4ccd3.

File: src/definitions/modules/flyout.less
Comment (prudho) [reply src/definitions/modules/flyout.less]: Resolved by deb1a561dfc99c26494d0e5c503c1118c7d4ccd3.

File: src/themes/default/globals/variation.variables
Comment (prudho) [reply src/themes/default/globals/variation.variables]: Resolved by deb1a561dfc99c26494d0e5c503c1118c7d4ccd3.

File: src/definitions/modules/flyout.js
Comment (lubber-de) [reply src/definitions/modules/flyout.js:70]: @prudho Change this to
```suggestion
        $context        = [window,document].indexOf(settings.context) < 0 ? $(document).find(settings.context) : $(settings.context),
```

Comment (prudho) [reply src/definitions/modules/flyout.js]: Fixed by 1d88605caabd4e4e7fab3a66a95ec0ae996636a3.

File: src/definitions/modules/flyout.js
Comment (prudho) [reply src/definitions/modules/flyout.js]: Fixed by1d88605caabd4e4e7fab3a66a95ec0ae996636a3.

**Comment by prudho** [comment]
@fomantic/maintainers I think Flyout module is ready for release now. Code has been cleaned and everything looks good from my POV. Many thanks to @lubber-de from helping me about the CSS stuff that I don't rule yet 😄 and @jamessampford for the initial code and idea.

For now module is pretty simple, but incremental versions of fomantic coupled with user feedback will help us to find the right direction of what we should do to improve it !

Documentation writing has started. If Docpad is kind enough to let me finish without crashing, it should be ready for the end of the week.

**Review by lubber-de — APPROVED** [review]
LGTM. I think i'll provide some enhancements to it (the "missing" features from modal like a11y support i mentioned in the other comments) _after_ this is merged.
However, i would like to wait for the your docs page before merging this 😉 Make sure to mark it as "beta" as we did with all other new components in the past.

**Comment by prudho** [comment]
Some issues to address before merging:
- Some events (here the `onHide` event) happens more and more time at each occurence.
capture

- Despite setting `scrollLock` to `true`, scrolling is still enabled on the dimmed page.
capture

**Comment by lubber-de** [comment]
> Despite setting scrollLock to true, scrolling is still enabled on the dimmed page.

I fixed this for sidebar by adopting the same logic from modal in #2436 , so flyout could simply adopt the same i believe :)

**Comment by prudho** [comment]
@lubber-de thanks for the scrollLock logic, now it works flawlessly :)

**Comment by lubber-de** [comment]
I am testing this again now, so we can merge it for 2.9.0 and enhance it in further patch releases 😉

**Review by lubber-de — APPROVED** [review]
LGTM

**Comment by lubber-de** [comment]
Docs added by https://github.com/fomantic/Fomantic-UI-Docs/pull/361

**Pull request merged**
